connections/sim/hw_sim.py

```diff
@@ -0,0 +1,89 @@
+from typing import Iterable, Tuple
+from itertools import repeat
+
+
+class IgnitorSim:
+    """
+    Simulates the hardware used for continuity checks.
+    """
+
+    # Corresponds to values returned by analogRead.
+    CONNECTED = 700  # Continuous
+    DISCONNECTED = 600  # Discontinuous
+    OFF = 0
+
+    def __init__(self, broken=False):
+        """
+        :param broken: If true, then it will simulate a used ignitor (and the continuity check should fail)
+        """
+        self._on = False
+        if broken:
+            self._on_level = self.DISCONNECTED
+        else:
+            self._on_level = self.CONNECTED
+
+    def write(self, val):
+        """
+        :brief: Write value to test pin, in manner of digitalWrite.
+        :param val: True (or castable to true) corresponds to a continuity test being done, and False indicates testing is complete.
+        """
+        if val:
+            self._on = True
+        else:
+            self._on = False
+
+    def read(self):
+        """
+        :brief: Simulate analogRead from output pin.
+        :return: Expected output.
+        """
+        if self._on:
+            return self._on_level
+        else:
+            return self.OFF
+
+    def fire(self):
+        self._on_level = self.DISCONNECTED
+
+
+class HWSim:
+    def __init__(
+        self, ignitors: Iterable[Tuple[int, int, int]] = tuple(), broken=False
+    ):
+        """
+        Implementation note: Default value for ignitors needs to be an immutable iterable.
+        :param ignitors: Iterable of 3-tuples containing test, read, and fire pin numbers.
+        In firmware's usage, to test continuity, the test pin is set high, and the voltage level at the read pin is used to determine whether the pin is continuous. To fire the pin, the fire pin is set high.
+        :param broken: If boolean, indicates whether all ignitors are broken. If iterable, should be the same length as ``ignitors``, and specify in order which ignitor is broken.
+        """
+        self.ignitors = []
+        self.ignitor_tests = {}
+        self.ignitor_reads = {}
+        self.ignitor_fires = {}
+        if isinstance(broken, bool):
+            broken = repeat(broken)
+        for (test, read, fire), broke in zip(ignitors, broken):
+            ign = IgnitorSim(broke)
+            self.ignitors.append(ign)
+            self.ignitor_tests[test] = ign
+            self.ignitor_reads[read] = ign
+            self.ignitor_fires[fire] = ign
+
+    def digital_write(self, pin, val):
+        """
+        :param pin: Should be a test pin.
```

Comment:
Should we maybe assert, throw, or warn if its not in any of the lists? Always good to fail noisily, might save us some day.

Reply:
Note that not all output pins are being simulated (yet) - e.g. some are connected to status LEDs, one is connected to a buzzer and is toggled at different frequencies to make different tones. IIRC I connected SIM directly to HW_Sim, so you'd have to define all of the valid output pins (as opposed to just the ones we're simulating).